### PR TITLE
Task/api 329 conformance api

### PR DESCRIPTION
**Pull request by bschofield-va**
http://hl7.org/fhir/DSTU2/conformance.html

### CODE REVIEW

File: argonaut-api/src/test/java/gov/va/api/health/argonaut/api/samples/SampleDataTypes.java [hunk 1]
```diff
@@ -85,10 +112,23 @@ public Narrative narrative() {
     return Narrative.builder().status(NarrativeStatus.additional).div("<p>HelloDiv<p>").build();
   }
 
+  public Period period() {
+    return Period.builder()
+        .id("5678")
+        .extension(singletonList(Extension.builder().url("http://wtf.com").valueInteger(1).build()))
```
Comment (jgarnhart-va) [inline argonaut-api/src/test/java/gov/va/api/health/argonaut/api/samples/SampleDataTypes.java]: I realize this is just a test, but maybe something other than `wtf`?

File: argonaut-api/src/main/java/gov/va/api/health/argonaut/api/ArgonautService.java [hunk 2]
```diff
@@ -68,6 +69,22 @@ Medication medicationRead(
...
 
+  @Operation(summary = "Conformance", description = "http://hl7.org/fhir/DSTU2/conformance.html")
+  @GET
+  @Path("metadata")
+  @ApiResponse(
+    responseCode = "200",
+    description = "Record found",
+    content =
+        @Content(
+          mediaType = "application/fhir+json",
```
[the remark this answers is not in the document]
Comment (jgarnhart-va) [reply argonaut-api/src/main/java/gov/va/api/health/argonaut/api/ArgonautService.java]: I recall us getting bit before for only supporting one way. for example "json+fhir" and not also "fhir+json". Just thought I'd pose the question here.

[the remark this answers is not in the document]
Comment (bschofield-va) [reply argonaut-api/src/main/java/gov/va/api/health/argonaut/api/ArgonautService.java]: Per the spec, the correct mime type is `application/json+fhir`. I've corrected the `ArgonautService` interface to ensure that is what is being published in the swagger document. 

The controller implementations are more forgiving and support 3 three variations.

```
@RequestMapping(
  value = {"/api/Patient"},
  produces = {"application/json", "application/json+fhir", "application/fhir+json"}
)
```
https://github.com/department-of-veterans-affairs/health-apis/blob/master/argonaut/src/main/java/gov/va/api/health/argonaut/service/controller/patient/PatientController.java#L42

http://hl7.org/fhir/DSTU2/http.html#mime-type

File: argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+  public static class RestSecurity implements BackboneElement {
+    @Pattern(regexp = Fhir.ID)
+    String id;
+
+    @Valid List<Extension> modifierExtension;
+    @Valid List<Extension> extension;
+    Boolean cors;
+    @Valid List<CodeableConcept> service;
+    String description;
+    @Valid SecurityCertificate certificate;
```
Comment (esubocz-va) [inline argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java]: Believe this should be a list of type SecurityCertificate (card is 0...*)

File: argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+    @Size(min = 1)
+    @Valid
+    List<ResourceInteraction> interaction;
+
+    RestResourceVersion versioning;
+    Boolean readHistory;
+    Boolean updateCreate;
+    Boolean conditionalCreate;
+    Boolean conditionalUpdate;
+    Boolean conditionalDelete;
```
Comment (esubocz-va) [inline argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java]: conditionalDelete should be of type code.

File: argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @AllArgsConstructor
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  public static class SearchParam implements BackboneElement {
+    @Pattern(regexp = Fhir.ID)
+    String id;
+
+    @Valid List<Extension> modifierExtension;
+    @Valid List<Extension> extension;
+    @NotEmpty String name;
```
Comment (esubocz-va) [inline argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java]: NotBlank recommended here.

Comment (esubocz-va) [reply argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java]: and may want to add @Valid to non-string types in this class. (type, modifier)

File: argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @AllArgsConstructor
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  public static class RestOperation implements BackboneElement {
+    @Pattern(regexp = Fhir.ID)
+    String id;
+
+    @Valid List<Extension> modifierExtension;
+    @Valid List<Extension> extension;
+    @NotEmpty String name;
```
Comment (esubocz-va) [inline argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java]: NotBlank recommended here.

File: argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+    @Pattern(regexp = Fhir.ID)
+    String id;
+
+    @Valid List<Extension> modifierExtension;
+    @Valid List<Extension> extension;
+    @NotNull @Valid Coding code;
+    MessagingEventCategory category;
+    @NotNull MessagingEventMode mode;
+
+    @NotEmpty
```
Comment (esubocz-va) [inline argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java]: NotBlank recommended here.
& @Valid recommended for category & mode,

File: argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+  @NoArgsConstructor(access = AccessLevel.PRIVATE)
+  @AllArgsConstructor
+  @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
+  public static class Document implements BackboneElement {
+    @Pattern(regexp = Fhir.ID)
+    String id;
+
+    @Valid List<Extension> modifierExtension;
+    @Valid List<Extension> extension;
+    @NotNull DocumentMode mode;
```
Comment (esubocz-va) [inline argonaut-api/src/main/java/gov/va/api/health/argonaut/api/resources/Conformance.java]: may want to add @Valid

**Pull request merged**